### PR TITLE
Bitstamp Support

### PR DESCRIPTION
**Pull request by michaelgrosner**
This is a partially completed implementation of the Bitstamp API for tribeca. I don't have a Bitstamp account, and I don't want to provide them with all the legal documents (https://www.bitstamp.net/account/verify/) for API access, so the authorized methods (buy, sell, account information, order status) are not tested. There is new config variables for Bitstamp and an Exchange enum value for Bitstamp now.

The market data API handling works just fine.

I'm leaving this pull request open until someone has the ability to implement the missing functionality and willing to test their implementation. I'm going to write some notes where I'm unsure of functionality.

### CODE REVIEW

File: src/service/gateways/bitstamp.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+	asks : [string, string][];
+}
+
+interface Ticker {
+	id : string;
+	amount : number;
+	price : number;
+}
+
+interface NewOrderAck {
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: I'm not sure if `price`, and `amount` are transmitted as number or strings. Based off their market data API, they're strings. We would need to convert from string to number if that is the case.

File: src/service/gateways/bitstamp.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    type: number; //  buy or sell (0 - buy; 1 - sell)
+    price: number;
+    amount: number;
+}
+
+interface Transaction {
+    datetime: string;
+    id: string;
+    type: number; //(0 - deposit; 1 - withdrawal; 2 - market trade)
+    usd: number;
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: Same as above - string or number?

File: src/service/gateways/bitstamp.ts [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    date: string;
+    tid: string;
+    price: string;
+    type: number;
+    amount: string;
+}
+
+interface AccountBalance {
+    usd_balance: number;
+    btc_balance: number;
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: Strings or numbers?

File: src/service/gateways/bitstamp.ts [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            })
+            .on("error", err => {
+                this.onConnectionStatusChange();
+                this._log("error %s", err);
+                throw err;
+            });
+	}
+}
+
+class BitstampMarketDataGateway implements Interfaces.IMarketDataGateway {
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: This class (and PusherClient) tested and works fine.

File: src/service/gateways/bitstamp.ts [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+            this._lastTimeMs++;
+        }
+        else {
+            this._lastTimeMs = t*100;
+        }
+        
+        return this._lastTimeMs;
+    };
+    
+    private addAuthentication = (req: {}) => {
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: Neither this nor the post method are not tested. I used https://github.com/askmike/bitstamp as a reference for the crypto functions.

File: src/service/gateways/bitstamp.ts [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+                orderStatus: Models.OrderStatus.Working
+            });
+        }).done();
+        return new Models.OrderGatewayActionReport(Utils.date());
+    };
+    
+    generateClientOrderId = () : string => shortId.generate();
+
+    private downloadOrderStatuses = () => {
+        // I really have no idea what to do here.
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: I don't know how to get order status updates for live orders from the Bitstamp API. I assume you need to use `order_status` but the docs are not super clear about what they return. I'm interested here if any orders were executed (partially or completely) well after being submitted.

File: src/service/gateways/bitstamp.ts [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        var defer = Q.defer<TResponse>();
+        request(options, (err, resp, body) => {
+            if (err) defer.reject(err);
+            else defer.resolve(JSON.parse(body));
+        });
+        return defer.promise;
+    };
+}
+
+class BitstampEntryGateway implements Interfaces.IOrderEntryGateway {
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: Nothing here was tested. I was just basing off the API documentation https://www.bitstamp.net/api/

File: src/service/gateways/bitstamp.ts [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+
+     _log : Utils.Logger = Utils.log("tribeca:gateway:BitstampOE");
+    
+    constructor(timeProvider : Utils.ITimeProvider, private _client : BitstampAuthenticatedClient) {
+        timeProvider.setInterval(this.downloadOrderStatuses, moment.duration(10, "seconds"));
+        timeProvider.setTimeout(() => this.ConnectChanged.trigger(Models.ConnectivityStatus.Connected), moment.duration(10));
+    }
+}
+
+class BitstampPositionGateway implements Interfaces.IPositionGateway {
```
Comment (michaelgrosner) [inline src/service/gateways/bitstamp.ts]: Again, not tested.

**Comment by hightower4586** [comment]
Would it be possible to reboot this? I would like to use it with Bitstamp and do have a verified account.

**Pull request closed without merging**